### PR TITLE
C#: Expression nullability

### PR DESCRIPTION
**Pull request by calumgrant**
- Extractor populates the nullability of expression types. `Expr.getAnnotatedType()` now works correctly.
- Sets the nullable flow state of each expression. `Expr.hasMaybeNullFlowState()` and `Expr.hasNotNullFlowState()` gives the nullable flow state as determined by Roslyn.

Merge https://git.semmle.com/Semmle/code/pull/35783 for corresponding DB hash changes.

### CODE REVIEW

File: csharp/ql/src/semmlecode.csharp.dbscheme [hunk 1]
```diff
@@ -483,7 +483,15 @@ case @nullability.kind of
 #keyset[parent, index]
 nullability_parent(int nullability: @nullability ref, int index: int ref, int parent: @nullability ref)
 
-type_nullability(int id: @has_type_annotation ref, int nullability: @nullability ref)
+type_nullability(int id: @has_type_annotation ref, int nullability: @nullability ref);
+
+/**
+ * Thw nullable flow state of an expression, as determined by Roslyn.
```
Comment (hvitved) [inline csharp/ql/src/semmlecode.csharp.dbscheme]: Typo, Thw -> The

**Comment by calumgrant** [comment]
Updated the tests and added nullness annotations to test expression nullability in the guards library. It turns out that one of the variables that had assumed not to be null guarded was in fact null guarded after all, which suggests that we are missing another null guard check somewhere.

**Comment by hvitved** [comment]
> Updated the tests and added nullness annotations to test expression nullability in the guards library. It turns out that one of the variables that had assumed not to be null guarded was in fact null guarded after all, which suggests that we are missing another null guard check somewhere.

I just checked, and I think something is dodgy with Roslyn's nullness analysis. For example, in
```
    string Test(object? o)
    {
        var t = o?.GetType();
        return t.ToString();
    }
```
it claims
- incorrectly that the call to `GetType()` is non-`null`,
- correctly that the call to `ToString()` is non-`null` (I wonder how, though, at it requires knowledge about the implementation of `ToString` on the `Type` class),
- nothing about the acces to `t` (even though it has concluded that the definition of `t` is non-`null`).

Perhaps what Roslyn means is that _if_ `o` and `t` are non-`null`, then so are the calls.

I think it is safer that we revert the 4 last commits (for now), and then reinvestigate in a follow up PR. Sorry about suggesting this addition in the first place, I thought it would have been straightforward.

**Comment by hvitved** [comment]
> I think it is safer that we revert the 4 last commits (for now), and then reinvestigate in a follow up PR. Sorry about suggesting this addition in the first place, I thought it would have been straightforward.

As we discussed offline already, even without the last 4 commits it is not safe to merge this PR, given that `hasNotNullFlowState()` is incorrect.

**Comment by calumgrant** [comment]
As agreed, I have removed the last 4 commits. I investigated the flow state of both the converted and unconverted types for the contentious cases but they were the same, so further investigation is required.

File: csharp/ql/test/library-tests/controlflow/graph/ExtractorOptions.cs [hunk 2]
```diff
@@ -1 +1,2 @@
 // semmle-extractor-options: /langversion:8.0
+#nullable enable
```
Comment (hvitved) [inline csharp/ql/test/library-tests/controlflow/graph/ExtractorOptions.cs]: Why this addition?

Comment (calumgrant) [reply csharp/ql/test/library-tests/controlflow/graph/ExtractorOptions.cs]: With this setting, unannotated reftypes are , andIt enables the feature in Roslyn. By default, flow state is not computed, and the types are assumed to be nullable.

Comment (calumgrant) [reply csharp/ql/test/library-tests/controlflow/graph/ExtractorOptions.cs]: With this setting, unannotated reftypes are non-nullable, and flow state is computed.  Without this settings, unannotated reftypes are nullable and flow state is not computed. 

We want to enable nullability for this test to make it interesting.

Comment (hvitved) [reply csharp/ql/test/library-tests/controlflow/graph/ExtractorOptions.cs]: Still, this is the only change to `library-tests/controlflow/graph`, I thought we only needed it for `library-tests/controlflow/guards` (where it is also added [here](https://github.com/Semmle/ql/pull/2450/files?file-filters%5B%5D=.cs#diff-3d93762a28c88302b14baf20534d06dfR2)).

Comment (calumgrant) [reply csharp/ql/test/library-tests/controlflow/graph/ExtractorOptions.cs]: It might be spurious -- let me check.

**Pull request merged**